src/libsync/discovery.cpp

```diff
@@ -486,11 +491,40 @@ void ProcessDirectoryJob::processFileAnalyzeRemoteInfo(
             } else {
                 item->_instruction = CSYNC_INSTRUCTION_SYNC;
             }
-        } else if (dbEntry._remotePerm != serverEntry.remotePerm || dbEntry._fileId != serverEntry.fileId) {
+        } else if (dbEntry._remotePerm != serverEntry.remotePerm || dbEntry._fileId != serverEntry.fileId || metaDataSizeNeedsUpdateForE2EeFilePlaceholder) {
+            if (metaDataSizeNeedsUpdateForE2EeFilePlaceholder) {
+                // we are updating placeholder sizes after migrating from older versions with VFS + E2EE implicit hydration not supported
+                qCInfo(lcDisco) << "Migrating the E2EE VFS placeholder " << dbEntry.path() << " from older version. The old size is " << item->_size << ". The new size is " << sizeOnServer;
+                item->_size = sizeOnServer;
+            }
             item->_instruction = CSYNC_INSTRUCTION_UPDATE_METADATA;
             item->_direction = SyncFileItem::Down;
         } else {
-            processFileAnalyzeLocalInfo(item, path, localEntry, serverEntry, dbEntry, ParentNotChanged);
+            // if (is virtual mode enabled and folder is encrypted - check if the size is the same as on the server and then - trigger server query
+            // to update a placeholder with corrected size (-16 Bytes)
+            // or, maybe, add a flag to the database - vfsE2eeSizeCorrected? if it is not set - subtract it from the placeholder's size and re-create/update a placeholder?
+            const QueryMode serverQueryMode = [this, &dbEntry, &serverEntry]() {
+                if (dbEntry.isDirectory() && dbEntry._isE2eEncrypted) {
+                    qint64 localFolderSize = 0;
+                    const auto listFilesCallback = [this, &localFolderSize](const OCC::SyncJournalFileRecord &record) {
+                        if (record.isFile()) {
+                            // add CommonConstants::e2EeTagSize so we will know the size of E2EE file on the server
+                            localFolderSize += record._fileSize + CommonConstants::e2EeTagSize;
+                        } else if (record.isVirtualFile()) {
+                            // just a virtual file, so, the size must contain CommonConstants::e2EeTagSize if it was not corrected already
+                            localFolderSize += record._fileSize;
+                        }
+                    };
+                    if (_discoveryData->_statedb->listFilesInPath(dbEntry.path().toUtf8(), listFilesCallback)
```

Comment:
this condition is really hard to understand
could you split the call to `listFilesInPath` from the check on its result ?
the reason why `localFolderSize == serverEntry.sizeOfFolder` change the return value to `NormalQuery` is not obvious
what is the logic behind this ?

Reply:
@mgallien This is explained in comments before `const QueryMode serverQueryMode`; But, I will explain it here again.

So, the thing is, on the server, we have the file's size stored including its 16 Bytes encryption tag (for encrypted files only!). But! When we want to use implicit hydration with VFS to deal with encrypted files - we simply can not include those 16 Bytes in the placeholder, because, we are not going to write them into the file anyway. This is because those 16 Bytes are not a part of a file. Those are appended when encrypting the file (I am not really a fan of this approach, what we could do, is store those in a separate field in the metadata on the server, but, it is what we have now).

Now, previously, if we have been using the E2EE already with the VFS and have created placeholders, those placeholders would then include those 16 Bytes that we do not want to include. Now, how do we migrate from the version not supporting E2EE + VFS implicit hydration, to the one that supports it? I have found one approach that always works, that is, compare the total size of all files in the folder on the server with the total size of placeholders of the same folder on the client. Then, if both sizes are identical, we can say for sure - we had those placeholders created with an older version before the implicit hydration of E2EE + VFS was merged. This means we must now re-create all placeholders in such a way that they won't include 16 Bytes of encryption tag. Normal query means - we will iterate through all files in the folder instead of skipping them as if they were already synced. Let me know if you got more questions.

So, in a nutshell. Let's say, we have an encrypted folder with 2 files. First file is of size 528 Bytes (where 16 Bytes - is an encryption tag). The second file is of a size of 1030 Bytes (with 16 Bytes of encryption tag). Now, if we have ever synced those files with VFS mode before, we will have had placeholders of 528 and 1030 Bytes created. Now, if we add those numbers 528 + 1030 = 1558 - we will have the size of the encrypted folder with 2 files on the server. If this size equals our local folder's placeholders size (528 + 1030 = 1558) - then we know for sure, our folder contains placeholders that contain extra 16 Bytes that we don't need for implicit hydration and this will not allow us to hydrate them implicitly. So, we will trigger the NormalQuery type of sync from the server so that we will be able to create placeholders of 512 and 1024 Bytes. Now, placeholders are of the correct size, and, if we hydrate them implicitly, we will not need to provide 16 Bytes of encryption tag as we never write it to the final file. We will fill them with 512 and 1024 Bytes of original data.